subprojects/base-services/src/main/java/org/gradle/internal/classloader/ClassLoaderUtils.java

```diff
@@ -16,17 +16,57 @@
 package org.gradle.internal.classloader;
 
 import org.gradle.api.Nullable;
+import org.gradle.internal.Cast;
 import org.gradle.internal.UncheckedException;
 import org.gradle.internal.concurrent.CompositeStoppable;
+import org.gradle.internal.reflect.JavaMethod;
+import org.gradle.internal.reflect.JavaReflectionUtil;
+import sun.misc.Unsafe;
 
 import java.io.IOException;
+import java.lang.reflect.Field;
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.net.URLConnection;
 
 public abstract class ClassLoaderUtils {
+
+    private static final Unsafe UNSAFE;
+
+    static {
+        try {
+            Field theUnsafe = Unsafe.class.getDeclaredField("theUnsafe");
+            theUnsafe.setAccessible(true);
+            UNSAFE = (Unsafe) theUnsafe.get(null);
+        } catch (NoSuchFieldException e) {
+            throw new RuntimeException(e);
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static final JavaMethod<ClassLoader, Package[]> GET_PACKAGES_METHOD;
+    private static final JavaMethod<ClassLoader, Package> GET_PACKAGE_METHOD;
+
+    static {
+        GET_PACKAGES_METHOD = getMethodWithFallback(Package[].class, new Class[0], "getDefinedPackages", "getPackages");
+        GET_PACKAGE_METHOD = getMethodWithFallback(Package.class, new Class[] {String.class}, "getDefinedPackage", "getPackage");
+    }
+
+    private static <T> JavaMethod<ClassLoader, T> getMethodWithFallback(Class<T> clazz, Class<?>[] params, String firstChoice, String fallback) {
+        JavaMethod<ClassLoader, T> method;
+        try {
+            method = JavaReflectionUtil.method(ClassLoader.class, clazz, firstChoice, params);
+        } catch (Throwable e) {
```

Comment:
Because Java sucks at exceptions handling :) I agree with you. Unfortunately, we don't really know which exceptions might be thrown here. Typically, `NoClassDefFoundError` is an unchecked exception, and it was impossible to know it would occur before actually seeing tests failing. As I am not sure what other exceptions might be thrown, possibly by other VM vendors implementations, I felt back to the old trick of using `Throwable`. Which, in this case, isn't a big deal since it's used in a static initializer for a class that is central to Gradle. In other words, odds of catching `OutOfMemoryException` are very low.